File: backend/routes/pos.py
"""
=============================================================================
 POS BOOTSTRAP ROUTE — routes/pos.py
=============================================================================

 API Aggregation endpoint that returns everything the POS screen needs
 in a single request.  Eliminates 5+ separate API calls on POS load.

 GET /api/pos/bootstrap
 →  products, categories, workers, settings, next_bill_number
=============================================================================
"""

from flask import Blueprint, jsonify
import cache
from services.db_service import DatabaseService
from models import db, Bill, Worker, Settings
from sqlalchemy import func
from datetime import date

pos_bp = Blueprint('pos', __name__, url_prefix='/api/pos')
db_service = DatabaseService()


@pos_bp.route('/bootstrap', methods=['GET'])
def bootstrap():
    """
    Single aggregated endpoint for POS initialization.
    
    Pulls from in-memory cache when available, falls back to DB.
    This replaces 5 separate API calls:
      GET /products
      GET /categories
      GET /workers
      GET /settings
      GET /bill/next-number
    """
    try:
        # 1. Products (with stock)
        products = cache.get('products_with_stock', 'active')
        if products is None:
            products = db_service.get_all_products_with_stock(include_inactive=False)
            cache.set('products_with_stock', 'active', products)

        # 2. Categories
        categories = cache.get('categories', 'active')
        if categories is None:
            categories = db_service.get_all_categories(include_inactive=False)
            cache.set('categories', 'active', categories)

        # 3. Workers
        workers = cache.get('workers', 'active')
        if workers is None:
            worker_objs = Worker.query.filter_by(status='active').all()
            workers = [{
                'worker_id': w.worker_id,
                'name': w.name,
                'role': w.role,
                'phone': w.phone,
                'status': w.status,
            } for w in worker_objs]
            cache.set('workers', 'active', workers)

        # 4. Settings
        settings = cache.get('settings', 'all')
        if settings is None:
            settings = db_service.get_all_settings()
            cache.set('settings', 'all', settings)

        # 5. Next bill number (always fresh — lightweight query)
        today = date.today()
        max_bill = db.session.query(func.max(Bill.bill_no)).filter(
            func.date(Bill.created_at) == today
        ).scalar()
        next_bill_number = (max_bill or 0) + 1

        return jsonify({
            'success': True,
            'products': products,
            'categories': categories,
            'workers': workers,
            'settings': settings,
            'next_bill_number': next_bill_number,
            '_cache_stats': cache.stats()  # For debugging during development
        })

    except Exception as e:
        return jsonify({
            'success': False,
            'message': f'Bootstrap failed: {str(e)}'
        }), 500

File: frontend/src/api/pos.js
import api from './api';

/**
 * POS Bootstrap API
 * 
 * Single endpoint that loads everything the POS screen needs in one request.
 * Replaces 5+ separate API calls with a single aggregated call.
 */
export const posAPI = {
    /**
     * Bootstrap POS — load all required data in one shot.
     * 
     * Response:
     * {
     *   products: [...],        // active products with stock
     *   categories: [...],      // active categories
     *   workers: [...],         // active workers
     *   settings: {...},        // all settings
     *   next_bill_number: 42    // next bill number
     * }
     */
    bootstrap: async () => {
        const response = await api.get('/api/pos/bootstrap');
        return response.data;
    },

    /**
     * Lightweight refresh — only products with stock levels.
     * Called after bill creation to update stock badges without
     * reloading the entire bootstrap payload.
     */
    refreshProducts: async () => {
        const response = await api.get('/api/products?include_stock=true');
        return response.data;
    },
};

export default posAPI;
